### PR TITLE
Fix the tsc_deadline issue when lapic_pt

### PR DESCRIPTION
**Pull request by lyan3**
Two changes are included:
- add "invariant TSC" capability detection at boot time, panic if the feature is not supported;
- fix tsc_deadline correctness issue by trapping TSC_DEADLINE msr write if VMX_TSC_OFFSET is not 0.

Tracked-On: #3636
Signed-off-by: Yan, Like <like.yan@intel.com>
Acked-by: Eddie Dong <eddie.dong@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
out of license

File: hypervisor/arch/x86/guest/vmsr.c [hunk 1]
```diff
@@ -561,6 +580,8 @@ static void set_guest_tsc_adjust(struct acrn_vcpu *vcpu, uint64_t tsc_adjust)
 
 	/* IA32_TSC_ADJUST is supposed to carry the value it's written to */
 	vcpu_set_guest_msr(vcpu, MSR_IA32_TSC_ADJUST, tsc_adjust);
+
+	set_tsc_msr_intercept(vcpu, (tsc_offset + tsc_adjust_delta ) != 0U);
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vmsr.c]: 0UL

File: hypervisor/arch/x86/guest/vmsr.c [hunk 2]
```diff
@@ -788,5 +809,5 @@ void update_msr_bitmap_x2apic_passthru(struct acrn_vcpu *vcpu)
 	enable_msr_interception(msr_bitmap, MSR_IA32_EXT_XAPICID, INTERCEPT_READ);
 	enable_msr_interception(msr_bitmap, MSR_IA32_EXT_APIC_LDR, INTERCEPT_READ);
 	enable_msr_interception(msr_bitmap, MSR_IA32_EXT_APIC_ICR, INTERCEPT_WRITE);
-	enable_msr_interception(msr_bitmap, MSR_IA32_TSC_DEADLINE, INTERCEPT_DISABLE);
+	set_tsc_msr_intercept(vcpu, exec_vmread64(VMX_TSC_OFFSET_FULL) != 0U);
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vmsr.c]: 0UL

File: hypervisor/arch/x86/guest/vmsr.c [hunk 3]
```diff
@@ -514,6 +531,8 @@ static void set_guest_tsc(struct acrn_vcpu *vcpu, uint64_t guest_tsc)
 
 	/* write to VMCS because rdtsc and rdtscp are not intercepted */
 	exec_vmwrite64(VMX_TSC_OFFSET_FULL, tsc_delta);
+
+	set_tsc_msr_intercept(vcpu, tsc_delta != 0U);
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vmsr.c]: 0UL?

File: hypervisor/arch/x86/guest/vmsr.c (hunk 1 above)
Comment (lyan3) [reply hypervisor/arch/x86/guest/vmsr.c]: Fixed, thanks.

**Comment by acrnsi** [comment]
OK to verify

**Comment by HuiHuangShi** [comment]
out of license

**Comment by HuiHuangShi** [comment]
14 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/arch/x86/cpu_caps.c func:detect_hardware_support offset:362<br> reason:Cyclomatic complexity greater than 20.<br><2> config:sharing_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:492<br> reason:Pointer param should be declared pointer to const. : vcpu<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:492<br> reason:Pointer param should be declared pointer to const. : vcpu<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:498<br> reason:Pointer not checked for null before use. : msr_bitmap<br><5> config:sharing_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:498<br> reason:Pointer not checked for null before use. : msr_bitmap<br><6> config:sharing_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:501<br> reason:Pointer not checked for null before use. : msr_bitmap<br><7> config:part_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:501<br> reason:Pointer not checked for null before use. : msr_bitmap<br><8> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:set_guest_tsc offset:535<br> reason:Implicit conversion: actual to formal param (MR). :  (bool and struct acrn_vcpu):  param  1<br><9> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:set_guest_tsc_adjust offset:584<br> reason:Implicit conversion: actual to formal param (MR). :  (unsigned long and bool):  param  2<br><10> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:set_guest_tsc_adjust offset:584<br> reason:Implicit conversion: actual to formal param (MR). :  (bool and struct acrn_vcpu):  param  1<br><11> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:update_msr_bitmap_x2apic_passthru offset:812<br> reason:Value is not of appropriate type. :  (int and unsigned long): exec_vmread64 ( 0x00002010U )   != 0UL<br><12> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:update_msr_bitmap_x2apic_passthru offset:812<br> reason:Implicit conversion: actual to formal param (MR). :  (bool and struct acrn_vcpu):  param  1<br><13> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:update_msr_bitmap_x2apic_passthru offset:812<br> reason:Implicit conversion: actual to formal param (MR). :  (int and bool):  param  2<br><14> config:common filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_set_tsc_deadline_msr offset:470<br> reason:Else alternative missing in if.<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by HuiHuangShi** [comment]
13 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/arch/x86/cpu_caps.c func:detect_hardware_support offset:362<br> reason:Cyclomatic complexity greater than 20.<br><2> config:sharing_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:492<br> reason:Pointer param should be declared pointer to const. : vcpu<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:492<br> reason:Pointer param should be declared pointer to const. : vcpu<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:498<br> reason:Pointer not checked for null before use. : msr_bitmap<br><5> config:sharing_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:498<br> reason:Pointer not checked for null before use. : msr_bitmap<br><6> config:sharing_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:501<br> reason:Pointer not checked for null before use. : msr_bitmap<br><7> config:part_config filename:/hypervisor/arch/x86/guest/vmsr.c func:set_tsc_msr_intercept offset:501<br> reason:Pointer not checked for null before use. : msr_bitmap<br><8> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:set_guest_tsc offset:535<br> reason:Implicit conversion: actual to formal param (MR). :  (bool and struct acrn_vcpu):  param  1<br><9> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:set_guest_tsc_adjust offset:584<br> reason:Implicit conversion: actual to formal param (MR). :  (unsigned long and bool):  param  2<br><10> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:set_guest_tsc_adjust offset:584<br> reason:Implicit conversion: actual to formal param (MR). :  (bool and struct acrn_vcpu):  param  1<br><11> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:update_msr_bitmap_x2apic_passthru offset:812<br> reason:Value is not of appropriate type. :  (int and unsigned long): exec_vmread64 ( 0x00002010U )   != 0UL<br><12> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:update_msr_bitmap_x2apic_passthru offset:812<br> reason:Implicit conversion: actual to formal param (MR). :  (bool and struct acrn_vcpu):  param  1<br><13> config:common filename:/hypervisor/arch/x86/guest/vmsr.c func:update_msr_bitmap_x2apic_passthru offset:812<br> reason:Implicit conversion: actual to formal param (MR). :  (int and bool):  param  2<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Pull request merged**